### ISSUE TITLE
rebase to absolute relative to root fails since 2.1.2

### ISSUE THREAD

**Issue by jybleau** [scniro/gulp-clean-css#22]
Related to this modification [(commit 3104238, line 33 to 38)](https://github.com/scniro/gulp-clean-css/commit/3104238c01fb0ebc5b32c0d61cf13788f9339b4d#diff-168726dbe96b3ce427e7fedce31bb0bcL33)

I'm rebasing relative paths in css files located in `bower_components` package folders.

For instance, an url in `font-awesome` package might look like:
`url(../fonts/fontawesome-webfont.eot?v=4.7.0);`

It is then rebased to its absolute path relative to the app's root, like this:
`url(/bower_components/font-awesome/fonts/fontawesome-webfont.eot?v=4.7.0);`

Now that the code on lines [33..38] is removed, the output is not the same when using theses options:

```
.pipe(cleanCSS({
      rebase: true,
      target: './',
      root: './'
    }))
```

The output looks like this instead:
`url(/Users/jybleau/Bitbucket/fonts/fontawesome-webfont.eot?v=4.7.0)`

Seems `cssFile[file.path]` is used by clean-css to know the complete absolute path to the css file being processed.
In the example above, it's missing the part there: `/Bitbucket/..??../fonts/`

If I paste back lines [33..38] in version 2.1.2+, it works fine again.

Thanks

PS: the reason for this rebasing is because all "vendors" css are merged together but the files they refer to are left in `bower_components` and used from there. It's a way of not having to copy all fonts and images files from each packages to a common location.

**Comment by jybleau** [comment]
...looks like it contradicts #8 
Ay ay...

**Comment by scniro** [comment]
sigh, so the pattern continues of fixing one thing and breaking another. Any suggestions?

**Comment by jybleau** [comment]
...trying to figure that out...

I'm not sure about the #8 issue, I'll give it a test; clean-css doc says: "Use a combination of relativeTo and target options for relative rebase". 

I don't see at any moment that he used both options.

Anyway, the tests will tell.

**Comment by madwizard-thomas** [comment]
I only used the target option in my own code, but gulp-clean-css [forces](https://github.com/scniro/gulp-clean-css/blob/master/index.js#L28) the relativeTo option for every file, so it is set.
```javascript
   if (!fileOptions.relativeTo && (fileOptions.root || file.path))
      fileOptions.relativeTo = path.dirname(path.resolve(options.root || file.path));
```
> Seems cssFile[file.path] is used by clean-css to know the complete absolute path to the css file being processed.
When cssFile[file.path] is used instead of the source itself, clean-css will rebase the sourcefile while reading the file (see source-reader.js(https://github.com/jakubpawlowicz/clean-css/blob/ea58a25f8e3edc206e902449a454175db85b6235/lib/utils/source-reader.js#L62-L80)), to either target or the root setting. This way if you specify multiple paths on the command line of clean-css, it will rebase the files as if they were in the same directory, then process them from there. 

However relativeTo (the path where the file is supposed to be rebased from) is set to the file path itself by gulp-clean-css. If the cssFile[file.path] method is used, the file will already be rebased and the relativeTo path is actually wrong. clean-css will rebase the file a second time this way.

So either the source should be supplied as a string with relativeTo set correctly, or the cssFile[file.path] method can be used but then relativeTo should probably be set to something else. 

I think in your case the problem might be that the root option is used as relativeTo, and I think this should always be the path to the actual file. Otherwise clean-css has no idea where the file is since after the breaking commit this is currently the only place where the filename is passed to clean-css.

It would be useful to have a test set with all the different cases so that breaking changes can be tested for easier.

**Comment by madwizard-thomas** [comment]
Come to think of it, it might be more true to clean-css to keep the cssFile[file.path] method but then relativeTo should not be set to anything by gulp-clean-css (but just implictly copied from options.relativeTo). After all in clean-css, relativeTo is a global option as well, not per file.

**Comment by jybleau** [comment]
Thanks @madwizard-thomas, it's getting clearer.
I will have time to look at it next week. 

Yup, let's add more test cases...

**Comment by madwizard-thomas** [comment]
For your information, the rebase algorithm in clean-css itself works like this (given options.rebase = true, which is the default)

Only when css[filePath] is used, for every input file:
Rebase from the directory of CSS source file
Rebase to (uses first one of these that is set):
1. **options.target** (relative rebase)
2. **options.root** (absolute rebase)
3. **process.cwd()** (relative rebase)

After that, regardless of the method used, the code is minified and then rebased:
- If **options.relativeTo** is not set, **no rebase** is done
- If **options.root** is set, absolute rebase from **relativeTo** to options.root as / 
- Else, if **options.target*** is set, relative rebase from **relativeTo** to **target**

(hopefully I didn't make any mistakes here. There is also a rebase for import statements but I haven't looked into that).

**Comment by scniro** [comment]
@madwizard-thomas thanks so much for really taking a deep dive into this. In your case that we fixed over in #8, if we change the following to your proposal, are you still issue-free? Here's what I am thinking...

```
fileOptions.relativeTo = options.relativeTo || file.path;

// back to path-keyed object
var cssFile = {};
cssFile[file.path] = {styles: file.contents ? file.contents.toString() : ''};

new CleanCSS(fileOptions).minify(cssFile, function (errors, css) {
//[...]
```

[here is a full gist](https://gist.github.com/scniro/8b1ce9f8c3918c33f5baa2b6c3da9da7) for index.js. Could you try pasting that into your `node_modules` installation and let me know if it works per your use case?

FWIW, this all may bubble up again once 4.0 drops. Check out the early proposed changes per rebasing jakubpawlowicz/clean-css#842

(context)
> code in the current master gets rid of root, relativeTo, target options in favor of just a single option rebaseTo

**Comment by madwizard-thomas** [comment]
I decided to make a test set with several cases, I've created a pull request (#24) for it. The gist you linked doesn't seem to work in all cases but the code in my pull request seems to work correctly. Needs to be checked though. I've tried to mimic the behavior of the CLI cleancss as much as possible.

**Comment by scniro** [comment]
merged! 2.3.1 is up, all looks good. Thank you so much for your help on this one. @jybleau could you install latest and let us know if you're squared away?

**Comment by scniro** [comment]
Also, I published 2.3.2 just now as there was a [clean-css release ](https://github.com/jakubpawlowicz/clean-css/releases/tag/v3.4.23) very recently that I wanted to get in

**Comment by jybleau** [comment]
All working fine now. Thanks so much for your work.
